python/sglang/srt/layers/quantization/modelopt_quant.py

```diff
@@ -1191,6 +1227,21 @@ def apply(
             )[0]
             if routed_scaling_factor is not None:
                 output *= routed_scaling_factor
+
+            if should_use_flashinfer_cutlass_moe_fp4_allgather():
+                output, global_hidden_states = (
+                    forward_batch.gathered_buffer[: forward_batch.input_ids.shape[0]],
```

Comment:
The code logic is getting increasingly messy. It seems that `forward_batch` is only need for dp communication. We need to refactor `dp_attention.py` so that any part of the code can handle dp communication without using `forward_batch`.  Let me try this later.

Reply:
I will work on this: https://github.com/sgl-project/sglang/pull/9136